### PR TITLE
Make CIS-2 token lookup more lenient

### PR DESCRIPTION
**Pull request by shjortConcordium**
## Purpose

Make CIS-2 token lookup more lenient

## Changes

- Redo `getTokens` so that metadataURL and balances are fetched individually.
- Log errors from token lookup, don't show them in UI.
- Don't fail "Choose Contract" unless all tokens are missing either metadata or balance.
- `fetchTokensConfigure` now keeps looking until 1 valid token is found (/or all tokens in the contract have been checked)

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Comment by orhoj** [comment]
The balance shows incorrectly when adding a token (it shows as 0 even if I have tokens).

File: packages/browser-wallet/src/popup/store/token.ts [hunk 1]
```diff
@@ -136,14 +137,22 @@ const cbf = atomFamily<string, Atom<ContractBalances>>((identifier: string) => {
...
             if (tokenIds.length !== 0) {
-                const balances = await getContractBalances(
-                    client,
-                    BigInt(contractIndex),
-                    0n,
-                    tokenIds,
-                    accountAddress,
-                    (error) => set(addToastAtom, error)
-                );
+                const instanceInfo = await client.getInstanceInfo(ContractAddress.create(BigInt(contractIndex), 0n));
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/store/token.ts:140]:
```suggestion
                const instanceInfo = await client.getInstanceInfo(ContractAddress.create(BigInt(contractIndex)));
```

File: packages/browser-wallet/src/popup/store/token.ts [hunk 2]
```diff
@@ -136,14 +137,22 @@ const cbf = atomFamily<string, Atom<ContractBalances>>((identifier: string) => {
...
+
+                let balances = {};
+                if (instanceInfo !== undefined) {
+                    balances = await getContractBalances(
+                        client,
+                        {
+                            index: BigInt(contractIndex),
+                            subindex: 0n,
+                            contractName: instanceInfo.name.value.substring(5),
```
Comment (orhoj) [inline packages/browser-wallet/src/popup/store/token.ts:149]: We have a helper function `getContractName` for this.

File: packages/browser-wallet/src/shared/utils/token-helpers.ts [hunk 3]
```diff
@@ -410,52 +405,42 @@ export async function getTokens(
...
-        contractDetails.index,
-        contractDetails.subindex,
-        ids,
-        account,
-        onError
+    const tokenData: (TokenData | undefined)[] = await Promise.all(
+        ids.map(async (id, index) => {
+            let metadataUrl;
+            try {
+                [metadataUrl] = await getTokenUrl(client, [id], contractDetails);
```
Comment (orhoj) [inline packages/browser-wallet/src/shared/utils/token-helpers.ts:412]: I think `getTokenUrl` is only ever invoked with a singleton list, so I would change the function to only support that.

**Review by orhoj — CHANGES_REQUESTED** [review]

File: packages/browser-wallet/src/popup/pages/Account/Tokens/ManageTokens/state.ts [hunk 4]
```diff
@@ -65,7 +65,7 @@ export const contractTokensAtom = (() => {
                     set(loadingAtom, true);
 
                     const tokens = get(listAtom);
-                    const topId = [...tokens].reverse()[0]?.pageId;
+                    const topId = tokens[tokens.length - 1]?.pageId;
```
[the remark this answers is not in the document]
Comment (shjortConcordium) [reply packages/browser-wallet/src/popup/pages/Account/Tokens/ManageTokens/state.ts:68]: The reason is to not reverse the list just to take the last element :thinking:

[the remark this answers is not in the document]
Comment (orhoj) [reply packages/browser-wallet/src/popup/pages/Account/Tokens/ManageTokens/state.ts:68]: Sure. I wonder why it was like that in the first place :smile:

File: packages/browser-wallet/src/shared/utils/token-helpers.ts (hunk 3 above)
Comment (shjortConcordium) [reply packages/browser-wallet/src/shared/utils/token-helpers.ts]: I just removed it, and used the helper from the SDK (did the same for getting balance)

**Comment by shjortConcordium** [comment]
> The balance shows incorrectly when adding a token (it shows as 0 even if I have tokens).

I'm unable to reproduce this? :thinking:

**Review by orhoj — APPROVED** [review]

**Pull request merged**